### PR TITLE
added dem generation code

### PR DESCRIPTION
**Pull request by Kelvinrr**
Added initial DEM generation code and blob change generation from DEM depressions and planform curvature. 

Kinda dumped things into utils for now, open to requests to move them elsewhere. 

Something to consider: Colins hill shade works pretty well but I'd be interested if there is a reasonable way to use [Gdal's hill shade utils](https://gdal.org/programs/gdaldem.html), wasn't able to find an easy python interface to it, though. 

example of output: 

topos

### CODE REVIEW

File: autocnet/utils/utils.py [hunk 1]
```diff
@@ -591,4 +597,289 @@ def import_func(func):
...
+        topo = topo + random.rand(int(nn), int(nn))*(200/(nn**alpha))
+    
+    topo = rd.rdarray(topo, no_data=0)
+    
+    curvatures = rd.TerrainAttribute(topo, attrib='slope_riserun')
+    curvatures = (curvatures - np.min(curvatures))/np.ptp(curvatures) * scale_factor
+    return topo - curvatures
+
+
+def hillshade(img, azi=255, min_slope=20, max_slope=100, min_bright=0, grayscale=False):
```
Comment (acpaquette) [inline autocnet/utils/utils.py:782]: This is missing the alt argument

File: autocnet/utils/utils.py [hunk 2]
```diff
@@ -591,4 +597,289 @@ def import_func(func):
...
+    hillshade a DEM, based on IDL code by Colin Dundas 
+    
+    Parameters
+    ----------
+    
+    img : np.array
+          DEM to hillshade
+    
+    azi : float 
+          Sun azimuth
```
Comment (acpaquette) [inline autocnet/utils/utils.py:793]: Can we mention that this should be in degrees?

File: autocnet/utils/utils.py [hunk 3]
```diff
@@ -591,4 +597,289 @@ def import_func(func):
...
     module = importlib.import_module(module, package='autocnet')
     func = getattr(module, func)
-    return func
\ No newline at end of file
+    return func
+
+
+def compute_depression(input_dem, scale_factor=1, curvature_percentile=75):
+    """
+    Compute depressions and return a new image with larges depressions filled in.
```
Comment (acpaquette) [inline autocnet/utils/utils.py:605]: larges to large

File: autocnet/utils/utils.py [hunk 4]
```diff
@@ -591,4 +597,289 @@ def import_func(func):
...
+
+    base_array = np.zeros(shape, dtype=dtype)  # Initialize your array of zeros
+
+    fill = np.ones(base_array.shape) * True  # Initialize boolean array defining shape fill
+
+    # Create check array for each edge segment, combine into fill array
+    for k in range(vertices.shape[0]):
+        fill = np.all([fill, check(vertices[k-1], vertices[k], base_array)], axis=0)
+    
+    print(fill.any())
```
Comment (acpaquette) [inline autocnet/utils/utils.py:734]: Remove print

File: autocnet/utils/utils.py [hunk 5]
```diff
@@ -591,4 +597,289 @@ def import_func(func):
...
+    for k in range(len(scales)):
+        nn = size/scales[k]
+        topo = scipy.misc.imresize(topo, (int(nn), int(nn)), "cubic", mode="F")
+        topo = topo + random.rand(int(nn), int(nn))*(200/(nn**alpha))
+    
+    topo = rd.rdarray(topo, no_data=0)
+    
+    curvatures = rd.TerrainAttribute(topo, attrib='slope_riserun')
+    curvatures = (curvatures - np.min(curvatures))/np.ptp(curvatures) * scale_factor
+    return topo - curvatures
```
Comment (acpaquette) [inline autocnet/utils/utils.py:779]: What's the curvature for here?

File: autocnet/utils/utils.py [hunk 6]
```diff
@@ -591,4 +597,289 @@ def import_func(func):
...
+    nx = (np.roll(dem, 1, axis = 1) - dem)
+    ny = (np.roll(dem, 1, axis = 0) - dem)
+    sz = np.shape(nx)
+    nz = np.ones(sz)
+    nl = np.sqrt(np.power(nx, 2.0) + np.power(ny, 2.0) + np.power(nz, 2.0))
+    nx = nx/nl
+    ny = ny/nl
+    nz = nz/nl
+
+    math.cos(math.radians(1))
```
Comment (acpaquette) [inline autocnet/utils/utils.py:845]: This can be removed, left over from prototyping

**Review by acpaquette — COMMENTED** [review]
Couple small things to change, and a small question regarding topology generation

File: autocnet/utils/utils.py (hunk 5 above)
Comment (Kelvinrr) [reply autocnet/utils/utils.py:779]: "Eroding" the cliff edges helps refine the topology a little creating something more organic, especially when using higher alphas.

**Pull request merged**